src/crypto/tls/handshake_messages.go

```diff
@@ -1340,6 +1367,14 @@ func marshalCertificate(b *cryptobyte.Builder, certificate Certificate) {
 						})
 					})
 				}
+				if certificate.DelegatedCredential != nil {
+					b.AddUint16(extensionDelegatedCredentials)
+					b.AddUint16LengthPrefixed(func(b *cryptobyte.Builder) {
+						b.AddUint24LengthPrefixed(func(b *cryptobyte.Builder) {
```

Comment:
I can't find where in the spec this Certificate extension is length-prefixed with 3 bytes -- can you point to it?

Reply:
Yeah, it does not; but I was thinking around if it should? There is no problem on the marshal side on this; but on the `unmarshalCertificate` function on handshake_messages.go, it:

```
		certificate.Certificate = append(certificate.Certificate, cert)
		for !extensions.Empty() {
			var extension uint16
			var extData cryptobyte.String
			if !extensions.ReadUint16(&extension) ||
				!extensions.ReadUint16LengthPrefixed(&extData) {
				return false
			}
			if len(certificate.Certificate) > 1 {
				// This library only supports OCSP and SCT for leaf certificates.
				continue
			}
```

expects it to be prefixed it with 16, so it will have that length prefixed... I don't know if that is ok...

I can also change completely `unmarshalCertificate`...

Reply:
It *seems* unnecessary with the 2-byte length prefix, right?

Reply:
Yeah.. I don't know why they always put it, maybe all the extensions they implemented so far had it...

Reply:
So,I've been thinking about this, and probably it will be good to have some kind of length, so it is actually known how much it is needed to read to have the DC.. what do you think?

Reply:
Doesn't the length of the extension give you that information?

Reply:
The 3 byte length prefix does not seem to belong here (see below for details). The value should be just the DelegatedCredential type.
___
Certificate Extensions use the "Extension" type per https://tools.ietf.org/html/rfc8446#section-4.4.2
```

      struct {
          /* ... */
          Extension extensions<0..2^16-1>;
      } CertificateEntry;
```

Extensions are encoded with a two-byte type, a 16-bit length field, and the actual value per https://tools.ietf.org/html/rfc8446#page-126
```
    struct {
        ExtensionType extension_type;
        opaque extension_data<0..2^16-1>;
    } Extension;
```

Per https://tlswg.org/tls-subcerts/draft-ietf-tls-subcerts.html#name-client-and-server-behavior
> The server MUST send the delegated credential as an extension in the CertificateEntry of its end-entity certificate

Per https://tlswg.org/tls-subcerts/draft-ietf-tls-subcerts.html#section-4
```
   struct {
     uint32 valid_time;
     SignatureScheme expected_cert_verify_algorithm;
     opaque ASN1_subjectPublicKeyInfo<1..2^24-1>;
   } Credential;
   struct {
     Credential cred;
     SignatureScheme algorithm;
     opaque signature<0..2^16-1>;
   } DelegatedCredential;
```

Reply:
Yes! Solved now!